### DISCUSSION TITLE
Workaround for providing support for System.Exception with System.Text.Json

### DISCUSSION THREAD

**Discussion by snjosephms** [dotnet/runtime#82299] [General]
We are migrating our code base from Newtonsoft to System.Text.Json. While testing this migration effort, we have encountered errors while working with System.Exception objects.

SerializedData Text: We see the following data in the serialized data created using System.Text.Json

{
    "Status": 0,
    "Exception": {
        "Message": **"TestExceptionMessag"**,
        "Data": {},
        "HResult": -2146233088
    },
    "SourceService": 1,
    "RequestId": "97f534a6-cfad-4d5f-934a-7298d27288a4",
    "Type": "ProvisionOC_Response",
    "RetryCount": 0,
    "IsRetryable": false,
    "ResubmitRetryCount": 0
}

**Issue**: 
The de-serialized data does not contain the Message text correctly. Message from deserialized object is **"Exception of type 'System.Exception' was thrown."**

Can you please share a workaround for providing support for System.Exception type with System.Text.Json?

**Comment by ghost** [comment]
Tagging subscribers to this area: @dotnet/area-system-text-json, @gregsdennis
See info in [area-owners.md](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) if you want to be subscribed.
<details>
<summary>Issue Details</summary>
<hr />

We are migrating our code base from Newtonsoft to System.Text.Json. While testing this migration effort, we have encountered errors while working with System.Exception objects.

SerializedData Text: We see the following data in the serialized data created using System.Text.Json

{
    "Status": 0,
    "Exception": {
        "Message": **"TestExceptionMessag"**,
        "Data": {},
        "HResult": -2146233088
    },
    "SourceService": 1,
    "RequestId": "97f534a6-cfad-4d5f-934a-7298d27288a4",
    "Type": "ProvisionOC_Response",
    "RetryCount": 0,
    "IsRetryable": false,
    "ResubmitRetryCount": 0
}

**Issue**: 
The de-serialized data does not contain the Message text correctly. Message from deserialized object is **"Exception of type 'System.Exception' was thrown."**

Can you please share a workaround for providing support for System.Exception type with System.Text.Json?

<table>
  <tr>
    <th align="left">Author:</th>
    <td>snjosephms</td>
  </tr>
  <tr>
    <th align="left">Assignees:</th>
    <td>-</td>
  </tr>
  <tr>
    <th align="left">Labels:</th>
    <td>

`area-System.Text.Json`

</td>
  </tr>
  <tr>
    <th align="left">Milestone:</th>
    <td>-</td>
  </tr>
</table>
</details>

**Comment by krwq** [comment]
You will need to write JsonConverter and factory for exceptions because Message doesn't have a setter (unless you're fine with setting that by reflection then you can also workaround it with contract model and providing Set for Message property). If exception is your own implementation then you can use JsonConstructor attribute and put it on the constructor which takes message

**Reply by snjosephms** [reply]
Exception is not our own implementation. We are using the  System.Exception class

**Comment by eiriktsarpalis** [comment]
@snjosephms could you share the expected serialization format?

**Reply by snjosephms** [reply]
Sorry, i didn't quite understand your question. The issue we see is that the exception message string we see on deserialization is not what was sent.

**Reply by eiriktsarpalis** [reply]
So you just want STJ to be able to roundtrip exceptions? Wouldn't it also need to be able to deserialize JSON created by Json.NET applications?

**Reply by snjosephms** [reply]
@eiriktsarpalis , the data we are trying to serialize contains a property of type System.Exception. 
The issue we see that on de-serialization, this property's  System.Exception.Message value is not getting retrieved correctly. I'm sorry if this was not clear from the query I posted.

**Reply by eiriktsarpalis** [reply]
I get that, my question is what JSON does Json.NET produce in your application? And given that you're in the process of migrating shouldn't the STJ converter be capable of deserializing that Json.NET payload apart from just roundtripping the `Message` property correctly?

**Reply by snjosephms** [reply]
Newtonsoft is results in the same Exception.Message that is being sent.
During this migration work, we are ensuring that STJ only deserializes json which is serialized using STJ.
Does that answer your question?

**Reply by eiriktsarpalis** [reply]
I understand that System.Text.Json does not roundtrip `Exception.Message` but Json.NET does. What I would need to know however is the JSON being produced by Json.NET currently.

**Reply by snjosephms** [reply]
Ok, I get your question now. I'll revert on this.

**Reply by snjosephms** [reply]
**Json produced by Newtonsoft**: "{\"Status\":\"NotStarted\",\"Exception\":{\"ClassName\":\"System.Exception\",\"Message\":\"TestExceptionMessage\",\"Data\":{\"$type\":\"System.Collections.ListDictionaryInternal, mscorlib\"},\"InnerException\":null,\"HelpURL\":null,\"StackTraceString\":null,\"RemoteStackTraceString\":null,\"RemoteStackIndex\":0,\"ExceptionMethod\":null,\"HResult\":-2146233088,\"Source\":null,\"WatsonBuckets\":null},\"SourceService\":\"ChannelConnector\",\"RequestId\":\"a4f8402f-d0b6-4f64-8d59-1f11e949ba5b\",\"Type\":\"PROVISIONOC_RESPONSE\",\"Version\":null,\"RetryCount\":0,\"IsRetryable\":false,\"ResubmitRetryCount\":0,\"OrgInfo\":null}"
**Json produced by STJ**: "{\"Status\":0,\"Exception\":{\"Message\":\"TestExceptionMessage\",\"Data\":{},\"HResult\":-2146233088},\"SourceService\":1,\"RequestId\":\"a4f8402f-d0b6-4f64-8d59-1f11e949ba5b\",\"Type\":\"ProvisionOC_Response\",\"RetryCount\":0,\"IsRetryable\":false,\"ResubmitRetryCount\":0}"

**Reply by snjosephms** [reply]
@eiriktsarpalis just wanted to follow up on this thread.

**Reply by eiriktsarpalis** [reply]
I see, so you're relying on your Exceptions' `ISerializable` output. STJ does not support `ISerializable` since it is associated with security flaws and is subject to deprecation in future release in .NET. That being said, it might be possible to write a custom converter for exceptions as follows:
```C#
public class ExceptionConverter : JsonConverter<Exception>
{
    public override bool CanConvert(Type type) => typeof(Exception).IsAssignableFrom(type);
    public override Exception Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        // Create a SerializationInfo object
        var info = new SerializationInfo(typeToConvert, new FormatterConverter());

        // Populate it with JSON properties
        while (reader.Read() && reader.TokenType != JsonTokenType.EndObject)
        {
            string propertyName = reader.GetString()!;
            reader.Read();
            object? value = JsonSerializer.Deserialize(ref reader, typeof(object), options);
            info.AddValue(propertyName, value);
        }

        // Use it to construct your object
        return (Exception)Activator.CreateInstance(typeToConvert, info, new StreamingContext())!;
    }

    public override void Write(Utf8JsonWriter writer, Exception value, JsonSerializerOptions options)
    {
        // Create a SerializationInfo object
        var info = new SerializationInfo(value.GetType(), new FormatterConverter());

        // Populate it with your object's fields
        value.GetObjectData(info, new StreamingContext());

        // Write it as JSON properties
        writer.WriteStartObject();

        // Write SerializationEntries as JSON properties
        foreach (SerializationEntry entry in info)
        {
            writer.WritePropertyName(entry.Name);
            JsonSerializer.Serialize(writer, entry.Value, options);
        }

        writer.WriteEndObject();
    }
}
```
